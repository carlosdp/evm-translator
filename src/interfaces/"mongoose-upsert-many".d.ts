declare module 'mongoose-upsert-many' {}
