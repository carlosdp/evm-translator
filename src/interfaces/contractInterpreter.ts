import { Address } from 'interfaces'

export type KeywordMap = {
    key: string
    filters: {
        [key: string]: string
    }
    defaultValue: string
    index?: number
}

export type MethodMap = {
    action: string
    exampleDescriptionTemplate: string
    exampleDescription: string
    keywords: Record<string, KeywordMap>
}

export type InterpreterMap = {
    contractAddress: Address
    methods: Record<string, string[]>
    contractOfficialName: string
    contractName: string
    writeFunctions: Record<string, MethodMap>
}
