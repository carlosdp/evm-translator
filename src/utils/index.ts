import { Chain } from '@type'

const ethereum: Chain = {
    EVM: true,
    id: 1,
    name: 'Ethereum',
    symbol: 'ETH',
    testnet: false,
    blockExplorerUrl: 'https://etherscan.io/',
}

export const chains: Record<string, Chain> = {
    ethereum,
}
