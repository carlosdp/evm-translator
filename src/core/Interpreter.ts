/* eslint-disable @typescript-eslint/no-non-null-assertion */
import contractInterpreters from './contractInterpreters'
import collect from 'collect.js'
import { BigNumber } from 'ethers'
import { formatEther } from 'ethers/lib/utils'
import {
    Address,
    Chain,
    Decoded,
    Interaction,
    InteractionEvent,
    Interpretation,
    RawTxData,
    Token,
    TokenType,
    TX_TYPE,
} from 'interfaces'
import { InterpreterMap, MethodMap } from 'interfaces/contractInterpreter'

function deepCopy(obj: any) {
    return JSON.parse(JSON.stringify(obj))
}

type ContractInterpretersMap = Record<Address, InterpreterMap>

type KeyMapping = {
    key: string
    defaultValue: string
    filters?: {
        event?: string
        to?: string
        from?: string
        value?: string
    } & Record<string, string>
    index?: number
    prefix?: string
    postfix?: string
}

const TopLevelInteractionKeys = ['contractName', 'contractSymbol', 'contractAddress', 'logIndex']
function isTopLevel(key: string) {
    return TopLevelInteractionKeys.includes(key)
}

// const getHardcodedContractInterpreters = (): any[] => {
//     let files: any[] = []
//     glob('**/contractInterpreters/*.json', (err, data) => {
//         console.log('ANYthing')
//         if (err) {
//             console.log('glob error:', err)
//         }
//         console.log('data', data)
//         const unique = data.slice(0, data.length / 2)
//         const filesNames = unique.map((item) => './' + item.split('/').slice(-2).join('/'))
//         console.log('filesNames', filesNames)
//         files = filesNames.map((fileName) => require(fileName)) as any[]
//         console.log('files', files[0])
//     })
//     return files
// }

class Interpreter {
    contractSpecificInterpreters: ContractInterpretersMap = {}
    // fallbackInterpreters: Array<Inspector> = []
    userAddress: Address
    chain: Chain

    constructor(userAddress: Address, chain: Chain) {
        this.userAddress = userAddress
        this.chain = chain

        for (const [address, map] of Object.entries(contractInterpreters)) {
            this.contractSpecificInterpreters[address as Address] = map as InterpreterMap
        }
    }

    public interpret(rawTxDataArr: RawTxData[], decodedDataArr: Decoded[]): Interpretation[] {
        const interpretations: Interpretation[] = []

        for (let i = 0; i < rawTxDataArr.length; i++) {
            const rawTxData = rawTxDataArr[i]
            const decodedData = decodedDataArr[i]

            const interpretation = this.interpretSingleTx(rawTxData, decodedData)
            interpretations.push(interpretation)
        }

        return interpretations
    }

    public interpretSingleTx(rawTxData: RawTxData, decodedData: Decoded): Interpretation {
        const method = decodedData.contractMethod!
        const interactions = decodedData.interactions!

        const toAddress = rawTxData.txReceipt.to
        const fromAddress = rawTxData.txReceipt.from

        const gasUsed = formatEther(
            BigNumber.from(decodedData.gasUsed).mul(BigNumber.from(decodedData.effectiveGasPrice)),
        )

        // not contract-specific
        const interpretation: Interpretation = {
            nativeTokenValueSent: decodedData.nativeTokenValueSent,
            tokensReceived: this.getTokensReiceived(interactions, this.userAddress),
            tokensSent: this.getTokensSent(interactions, this.userAddress),
            nativeTokenSymbol: this.chain.symbol,
            userName: decodedData.fromENS || fromAddress.substring(0, 6),
            gasPaid: gasUsed,
        }

        const interpretationMapping = this.contractSpecificInterpreters[toAddress]
        const methodSpecificMapping = interpretationMapping?.writeFunctions[method]

        // console.log('interpretationMapping', interpretationMapping)
        // contract-specific interpretation
        if (interpretationMapping && methodSpecificMapping) {
            // console.log('contract-specific interpretation', interpretationMapping.contractName, method)
            // some of these will be arribtrary keys
            interpretation.contractName = interpretationMapping.contractName
            interpretation.action = methodSpecificMapping.action
            interpretation.exampleDescription = methodSpecificMapping.exampleDescription

            if (decodedData.reverted) {
                interpretation.reverted = true
                // TODO the description and extras should be different for reverted transactions
            }
            interpretation.extra = this.useMethodMapping(
                interactions,
                methodSpecificMapping,
                interpretationMapping.contractAddress,
            )

            interpretation.exampleDescription = this.fillDescriptionTemplate(
                interpretationMapping.writeFunctions[method].exampleDescriptionTemplate,
                interpretation,
            )
        } else {
            // const fallbackData = this.fallbackInterpreters(rawTxData, decodedData)
        }

        return interpretation
    }

    // private genericERC721Interpreter(rawTxData: RawTxData, decodedData: Decoded): Record<string, any> {}

    private fallbackInterpreters(rawTxData: RawTxData, decodedData: Decoded): Record<string, any> {
        const data = {}

        // contract Deploy
        if (decodedData.txType === TX_TYPE.CONTRACT_DEPLOY) {
            return {}
        } else if (decodedData.txType === TX_TYPE.CONTRACT_INTERACTION) {
            return {}
        }

        return data
    }

    private findValue(
        interactions: Interaction[],
        keyMapping: KeyMapping,
        userAddress: Address,
        contractAddress: Address,
    ): string {
        const filters = keyMapping.filters || {}
        const index = keyMapping.index || 0

        let filteredInteractions = deepCopy(interactions) as Interaction[]

        // interactions.forEach((interaction) => {
        //     interaction.events.forEach((event) => {
        //         console.log('event', event)
        //     })
        // })

        for (const [key, value] of Object.entries(filters)) {
            let valueToFind = value

            // special case for when we want to match on the contextualized user address
            if (value === '{userAddress}') valueToFind = userAddress
            if (value === '{contractAddress}') valueToFind = contractAddress

            // filter out by events that don't have the keys we want
            for (const interaction of filteredInteractions) {
                if (!isTopLevel(key)) {
                    interaction.events = interaction.events.filter((d) => d[key] === valueToFind)
                }
            }

            // if there aren't any events left, we don't want that interaction
            filteredInteractions = filteredInteractions.filter((i) => i.events.length > 0)

            // filter out interactions by top level key
            if (isTopLevel(key)) {
                filteredInteractions = filteredInteractions.filter((i) => (i as any)[key] === valueToFind)
            }
        }

        const interaction = filteredInteractions[index]
        const prefix = keyMapping.prefix || ''
        const str = (interaction as any)?.[keyMapping.key] || interaction?.events[index]?.[keyMapping.key]
        const postfix = keyMapping.postfix || ''
        const value = str ? prefix + str + postfix : keyMapping.defaultValue

        return value
    }

    private getTokens(interactions: Interaction[], userAddress: Address, direction: 'to' | 'from'): Token[] {
        let tokens: Token[] = []
        type FlattenedInteraction = Omit<Interaction, 'events'> & InteractionEvent

        let filteredInteractions = deepCopy(interactions) as Interaction[]

        const toKeys = ['to', '_to']
        const fromKeys = ['from', '_from']
        const transferEvents = ['Transfer', 'TransferBatch', 'TransferSingle']

        for (const interaction of filteredInteractions) {
            // filter non-transfer events
            interaction.events = interaction.events.filter((d) => transferEvents.includes(d.event))
            // filter out events that aren't to/from the user in the right direction
            interaction.events = interaction.events.filter((d) => toOrFromUser(d, direction, userAddress))
        }

        // filter out interactions without any of the events we want
        filteredInteractions = filteredInteractions.filter(
            (i) =>
                i.events.filter(
                    (d) => transferEvents.includes(d.event), // filters out all non transfer events
                ).length > 0, // filters out interactions that don't have any events objects left
        )

        // console.log('filteredInteractions', filteredInteractions[0].events)
        const flattenedInteractions = flattenEvents(filteredInteractions)

        function getTokenType(interaction: FlattenedInteraction): TokenType {
            const LPTokenSymbols = ['UNI-V2']
            let tokenType = TokenType.DEFAULT

            // LP Token
            if (LPTokenSymbols.includes(interaction.contractSymbol)) {
                tokenType = TokenType.LPToken
                // ERC-1155
            } else if (interaction._amount || interaction._amounts) {
                tokenType = TokenType.ERC1155
                // ERC-721
            } else if (interaction.tokenId) {
                tokenType = TokenType.ERC721
                // ERC-20
            } else if (Number(interaction.value) > 0) {
                tokenType = TokenType.ERC20
            }

            return tokenType
        }

        function toOrFromUser(event: InteractionEvent, direction: 'to' | 'from', userAddress: Address) {
            const directionArr = direction === 'to' ? toKeys : fromKeys
            return directionArr.filter((key) => event[key] === userAddress).length > 0
        }

        // each transfer event can have a token in it. for example, minting multiple NFTs
        function flattenEvents(interactions: Interaction[]) {
            const flattenedInteractions: FlattenedInteraction[] = []

            let flattenedInteraction: FlattenedInteraction
            for (const interaction of interactions) {
                flattenedInteraction = deepCopy(interaction) as FlattenedInteraction
                delete flattenedInteraction.events

                for (const event of interaction.events) {
                    // flattenedInteraction.event = event.event
                    // flattenedInteraction.logIndex = event.logIndex
                    flattenedInteraction = { ...flattenedInteraction, ...event }
                    flattenedInteractions.push(flattenedInteraction)
                }
                // console.log('flattenedInteraction', flattenedInteraction)
            }

            return flattenedInteractions
        }

        tokens = flattenedInteractions.map((i) => {
            const tokenType = getTokenType(i)

            return {
                type: tokenType,
                name: i.contractName,
                symbol: i.contractSymbol,
                address: i.contractAddress,
                amount: i.value,
                tokenId: i.tokenId as string,
            }
        })

        return tokens
    }

    private getTokensReiceived(interactions: Interaction[], userAddress: Address): Token[] {
        return this.getTokens(interactions, userAddress, 'to')
    }

    private getTokensSent(interactions: Interaction[], userAddress: Address): Token[] {
        return this.getTokens(interactions, userAddress, 'from')
    }

    private useMethodMapping(
        interactions: Interaction[],
        methodSpecificMapping: MethodMap,
        contractAddress: Address,
    ): Record<string, string> {
        const keywordsMap = methodSpecificMapping.keywords
        const keyValueMap: Record<string, string> = {}

        const ignoreKeys = ['action', 'contractName', 'exampleDescription']

        for (const [key, value] of Object.entries(keywordsMap).filter(([key]) => !ignoreKeys.includes(key))) {
            // if the value is a string, we can just use it
            if (typeof value === 'string') {
                keyValueMap[key] = value

                // some data requires searching for it
            } else if (typeof value === 'object') {
                keyValueMap[key] = this.findValue(interactions, value, this.userAddress, contractAddress)
            }
        }

        return keyValueMap
    }

    private fillDescriptionTemplate(template: string, interpretation: Interpretation): string {
        const merged = collect(interpretation.extra).merge(interpretation).all()

        for (const [key, value] of Object.entries(merged)) {
            if (typeof value === 'string') {
                template = template.replace(`{${key}}`, value)
            }
        }
        return template
    }
}

export default Interpreter
