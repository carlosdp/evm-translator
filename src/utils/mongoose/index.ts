import { ContractModel } from './models/contract'
import { connect } from 'mongoose'

import { ContractData } from 'interfaces'
import { ABI_Event, ABI_EventZ, ABI_ItemUnfiltered, ABI_Row, ABI_RowZ, ABI_Type } from 'interfaces/abi'

import { DatabaseInterface } from 'utils/DatabaseInterface'
import { ABI_RowModel } from 'utils/mongoose/models/abi'

export class MongooseDatabaseInterface extends DatabaseInterface {
    connectionString: string

    constructor(connectionString: string) {
        super()
        this.connectionString = connectionString
    }
    async connect() {
        // TODO
        // mongoose.connect('mongodb://user:pass@localhost:port/database', { autoIndex: false }); for PROD
        await connect(this.connectionString)
    }

    async getContractDataForManyContracts(contractAddresses: string[]): Promise<Record<string, ContractData | null>> {
        const contractMap: Record<string, ContractData | null> = {}
        for (let i = 0; i < contractAddresses.length; i++) {
            contractMap[contractAddresses[i]] = null
        }

        try {
            const modelData = await ContractModel.find({ address: { $in: contractAddresses } })
            const data = modelData.map((model) => model.toObject())

            for (let i = 0; i < contractAddresses.length; i++) {
                const contractAddress = contractAddresses[i]
                const contractData = data.find((contract) => contract.address === contractAddress)
                contractMap[contractAddress] = contractData || null
            }
        } catch (e) {
            console.log('get contract mongoose error')
            console.log(e)
            // return null
        }
        // return here instead of in the try, so that it still works if the db is down
        return contractMap
    }

    async addOrUpdateManyContractData(contractDataArr: ContractData[]): Promise<void> {
        try {
            // only way to do bulk upsert
            const { result } = await ContractModel.bulkWrite(
                contractDataArr.map((contract) => ({
                    updateOne: {
                        filter: { address: contract.address },
                        update: contract,
                        upsert: true,
                    },
                })),
            )

            console.log('contracts:')
            console.log(result)
        } catch (e) {
            console.log('contract mongoose error')
            console.log(e)
        }
    }

    async addOrUpdateManyABI(abiArr: ABI_Row[]): Promise<void> {
        // prob don'`t need these 3 lines but might it optimize the writes
        const uniqueABIsAsStrings = new Set<string>()
        abiArr.map((abi) => uniqueABIsAsStrings.add(JSON.stringify(abi)))
        const uniqueABIs = Array.from(uniqueABIsAsStrings).map((str) => JSON.parse(str))

        try {
            // only way to do bulk upsert
            const { result } = await ABI_RowModel.bulkWrite(
                uniqueABIs.map((abi) => ({
                    updateOne: {
                        filter: { hashedSignature: abi.hashedSignature, fullSignature: abi.fullSignature },
                        update: abi,
                        upsert: true,
                    },
                })),
            )

            console.log('ABIs:')
            console.log(result)
        } catch (e) {
            console.log('abi mongoose error')
            console.log(e)
        }
    }

    async getABIsForHexSignature(hexSignature: string): Promise<ABI_ItemUnfiltered[]> {
        const abiModels = await ABI_RowModel.find({ hashedSignature: hexSignature })
        const abis = abiModels.map((abi) => ABI_RowZ.parse(abi.toObject()))
        return abis.map((abi) => abi.abiJSON)
    }
    async getFirstABIForHexSignature(hexSignature: string): Promise<ABI_ItemUnfiltered | null> {
        const abis = (await this.getABIsForHexSignature(hexSignature)) || []
        return abis.length ? abis[0] : null
    }

    async getEventABIsForHexSignature(hexSignature: string): Promise<ABI_Event[]> {
        const abiModels = await ABI_RowModel.find({ hashedSignature: hexSignature })
        const abisRows = abiModels.map((abi) => ABI_RowZ.parse(abi.toObject()))
        abisRows.sort((a, b) => Number(b.default || false) - Number(a.default || false)) // get default(s) first
        const abis = abisRows.map((abi) => abi.abiJSON)
        const events = abis.filter((abi) => abi.type === ABI_Type.enum.event)
        return events.map((event) => ABI_EventZ.parse(event))
    }
}
