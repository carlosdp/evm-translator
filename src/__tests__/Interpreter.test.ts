import Interpreter from '../core/Interpreter'
import { chains } from '../index'
import testTxHashes, { flattenTxHashes } from '../__tests__/testTxHashes.js'
import { MongooseDatabaseInterface } from 'utils/mongoose'
import { DecodedTx } from '../interfaces/decoded'

jest.mock('node-fetch', () => jest.fn())

jest.useRealTimers();

test('Interpreter', async () => {
  jest.setTimeout(20 * 1000);
  const db = new MongooseDatabaseInterface('mongodb://localhost:27017/evm-translator')
  await db.connect()
  const interpreter = new Interpreter(chains.ethereum)
  const txMap = await db.getManyDecodedTxMap(flattenTxHashes())
  const filteredDecodedTxes = Object.entries(txMap).filter(([txHash, decodedTx]) => {
    if (!decodedTx) {
      console.log('Issue with decoder for tx:', txHash)
      return false
    }
    return true
  }).map(([, decodedTx]) => (decodedTx))

  for (const decodedTx of filteredDecodedTxes) {
    expect(interpreter.interpretSingleTx(decodedTx!)).toMatchSnapshot()
  }
})
