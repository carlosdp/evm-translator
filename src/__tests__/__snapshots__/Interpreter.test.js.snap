// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interpreter 1`] = `
Object {
  "action": "sold",
  "chainSymbol": "ETH",
  "contractAddress": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
  "contractName": "Uniswap V2",
  "counterpartyName": null,
  "exampleDescription": "0xdcde sold 71782049366433048 ETH of KINGDOM on Uniswap V2",
  "extra": Object {
    "ethValueReceived": "71782049366433048",
    "tokenAmount": "458161051962122",
    "tokenName": "KINGDOM",
  },
  "gasPaid": "0.03127220866447174",
  "nativeValueReceived": "0.3949571157926511",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1638316845,
  "tokensReceived": Array [],
  "tokensSent": Array [
    Object {
      "address": "0x9273b71cfd60e022c2107d5bcf5ee62a88346550",
      "amount": "458161051962122",
      "name": "Kingdom Token",
      "symbol": "KINGDOM",
      "type": "ERC20",
    },
  ],
  "txHash": "0xd7be990f8849ec0b74d86fbda86eaf4c0cb25b814fb1ab9d2f056705b2fdfba6",
  "userAddress": "0xdcdee85472e19a9a1bba1b9c524c6105c727f3d6",
  "userName": "0xdcde",
}
`;

exports[`Interpreter 2`] = `
Object {
  "action": "bought",
  "chainSymbol": "ETH",
  "contractAddress": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
  "contractName": "Uniswap V2",
  "counterpartyName": null,
  "exampleDescription": "thememelord.eth bought 1 ETH of HOGE on Uniswap V2",
  "extra": Object {
    "tokenAmount": "11567525892826453",
    "tokenName": "HOGE",
  },
  "gasPaid": "0.023564857708634499",
  "nativeValueReceived": "0",
  "nativeValueSent": "1",
  "reverted": null,
  "timestamp": 1638316845,
  "tokensReceived": Array [
    Object {
      "address": "0xfad45e47083e4607302aa43c65fb3106f1cd7607",
      "amount": "11567525892826453",
      "name": "hoge.finance",
      "symbol": "HOGE",
      "type": "ERC20",
    },
  ],
  "tokensSent": Array [],
  "txHash": "0x32706fd90e141319710989feb555d6c9e394b1788802df6281d257cda0701c84",
  "userAddress": "0x10fced9a7c5d1ae7186d5e5912b73d47c1283d65",
  "userName": "thememelord.eth",
}
`;

exports[`Interpreter 3`] = `
Object {
  "action": "swapped",
  "chainSymbol": "ETH",
  "contractAddress": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
  "contractName": "Uniswap V2",
  "counterpartyName": null,
  "exampleDescription": "0x4090 swapped 900000000 USDC for 202899041047771 KINGDOM on Uniswap V2",
  "extra": Object {
    "tokenAmount0": "900000000",
    "tokenAmount1": "202899041047771",
    "tokenName0": "USDC",
    "tokenName1": "KINGDOM",
  },
  "gasPaid": "0.026273653630187205",
  "nativeValueReceived": "0",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1638316845,
  "tokensReceived": Array [
    Object {
      "address": "0x9273b71cfd60e022c2107d5bcf5ee62a88346550",
      "amount": "202899041047771",
      "name": "Kingdom Token",
      "symbol": "KINGDOM",
      "type": "ERC20",
    },
  ],
  "tokensSent": Array [
    Object {
      "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
      "amount": "900000000",
      "name": "USD Coin",
      "symbol": "USDC",
      "type": "ERC20",
    },
  ],
  "txHash": "0xb684992cdec31d2eeaf3866be82ddd2a9cfe2211bab96b1aa132d3a9ae50587b",
  "userAddress": "0x4090bd947f70b9b8c9522faa0d2f7f9bbb4b2377",
  "userName": "0x4090",
}
`;

exports[`Interpreter 4`] = `
Object {
  "action": "added liquidity",
  "chainSymbol": "ETH",
  "contractAddress": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
  "contractName": "Uniswap V2",
  "counterpartyName": null,
  "exampleDescription": "0x8b7b added liquidity to the ACYC-ETH Uniswap V2 pool",
  "extra": Object {
    "token_0": "ACYC",
  },
  "gasPaid": "0.037819919756787315",
  "nativeValueReceived": "0",
  "nativeValueSent": "10",
  "reverted": null,
  "timestamp": 1638316877,
  "tokensReceived": Array [
    Object {
      "address": "0xc7b7dac697b03e11e4af085d927f0c9b24b3f27d",
      "amount": "3162277660168379331997893",
      "name": "Uniswap V2",
      "symbol": "UNI-V2",
      "type": "LPToken",
    },
  ],
  "tokensSent": Array [
    Object {
      "address": "0xb56a1f3310578f23120182fb2e58c087efe6e147",
      "amount": "1000000000000000000000000000000",
      "name": "AllCoinsYieldCapital",
      "symbol": "ACYC",
      "type": "ERC20",
    },
  ],
  "txHash": "0xb78ce1e0f55e78cf005b4b5af9978b3d20292dc8c88c94f7ecade67083e2b97f",
  "userAddress": "0x8b7b8448243559a5daa98a22cb6982796bce4271",
  "userName": "0x8b7b",
}
`;

exports[`Interpreter 5`] = `
Object {
  "action": "removed liquidity",
  "chainSymbol": "ETH",
  "contractAddress": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
  "contractName": "Uniswap V2",
  "counterpartyName": null,
  "exampleDescription": "0x888a removed liquidity from the KAKASHI-WETH Uniswap V2 pool",
  "extra": Object {
    "token_0": "KAKASHI",
    "token_1": "WETH",
  },
  "gasPaid": "0.029644983060867926",
  "nativeValueReceived": "0",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1638318188,
  "tokensReceived": Array [
    Object {
      "address": "0xb7b284c2ea09c637daa2f128a8756280a9bd9091",
      "amount": "1868562427769924",
      "name": "Kakashi",
      "symbol": "KAKASHI",
      "type": "ERC20",
    },
    Object {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "amount": "4356835431801684676",
      "name": "Wrapped Ether",
      "symbol": "WETH",
      "type": "unknown",
    },
  ],
  "tokensSent": Array [
    Object {
      "address": "0xbe2cee7429e427c262e73ea074bef59c6ad1d93b",
      "amount": "93937438884835667",
      "name": "Uniswap V2",
      "symbol": "UNI-V2",
      "type": "LPToken",
    },
  ],
  "txHash": "0x967963d6eaad9e65d567cac325728d79aeaa319b1453e05ec55734f3b08445ec",
  "userAddress": "0x888ae876e588cd29715ab2b4a6ecce0bc2e44655",
  "userName": "0x888a",
}
`;

exports[`Interpreter 6`] = `
Object {
  "action": "removed liquidity",
  "chainSymbol": "ETH",
  "contractAddress": "0x7a250d5630b4cf539739df2c5dacb4c659f2488d",
  "contractName": "Uniswap V2",
  "counterpartyName": null,
  "exampleDescription": "0xc712 removed liquidity from the SAND-ETH Uniswap V2 pool",
  "extra": Object {
    "ethValueReceived": "6538029747179976387",
    "token_0": "SAND",
  },
  "gasPaid": "0.032858980114895286",
  "nativeValueReceived": "6.538029747179976",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1638318377,
  "tokensReceived": Array [
    Object {
      "address": "0x3845badade8e6dff049820680d1f14bd3903a5d0",
      "amount": "4456803721261159794400",
      "name": "SAND",
      "symbol": "SAND",
      "type": "ERC20",
    },
  ],
  "tokensSent": Array [
    Object {
      "address": "0x3dd49f67e9d5bc4c5e6634b3f70bfd9dc1b6bd74",
      "amount": "123932818358485202218",
      "name": "Uniswap V2",
      "symbol": "UNI-V2",
      "type": "LPToken",
    },
  ],
  "txHash": "0x3dd8f41c008a6ce59ca8d2b24df700dc4441d0a83747b9b7c0587449a2e7c9ca",
  "userAddress": "0xc712b31175906564bb41a7b84169bd4a9b522011",
  "userName": "0xc712",
}
`;

exports[`Interpreter 7`] = `
Object {
  "action": "bought",
  "chainSymbol": "ETH",
  "contractAddress": "0x7f268357a8c2552623316e2562d90e642bb538e5",
  "contractName": "OpenSea",
  "counterpartyName": null,
  "exampleDescription": "0x96a6 bought The Ninja Hideout #228 on OpenSea for 0.05 ETH",
  "extra": Object {
    "NFT": "The Ninja Hideout",
    "tokenId": "#228",
  },
  "gasPaid": "0.00837078350034528",
  "nativeValueReceived": "0",
  "nativeValueSent": "0.05",
  "reverted": null,
  "timestamp": 1656089719,
  "tokensReceived": Array [
    Object {
      "address": "0x97e41d5ce9c8cb1f83947ef82a86e345aed673f3",
      "name": "The Ninja Hideout",
      "symbol": "TNH",
      "tokenId": "228",
      "type": "ERC721",
    },
  ],
  "tokensSent": Array [],
  "txHash": "0xc814cb6b61222beda8a9bcc359e776e72ae732a1c29df572099ecef27c8461e4",
  "userAddress": "0x96a675775b8546a8602a1015efaacbc87358e478",
  "userName": "0x96a6",
}
`;

exports[`Interpreter 8`] = `
Object {
  "action": "sold",
  "chainSymbol": "ETH",
  "contractAddress": "0x7f268357a8c2552623316e2562d90e642bb538e5",
  "contractName": "OpenSea",
  "counterpartyName": null,
  "exampleDescription": "0x83eb sold The Ninja Hideout #228 on OpenSea for 0.04625 ETH",
  "extra": Object {
    "NFT": "The Ninja Hideout",
    "tokenId": "#228",
  },
  "gasPaid": "0.00837078350034528",
  "nativeValueReceived": "0.04625",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1656089719,
  "tokensReceived": Array [],
  "tokensSent": Array [
    Object {
      "address": "0x97e41d5ce9c8cb1f83947ef82a86e345aed673f3",
      "name": "The Ninja Hideout",
      "symbol": "TNH",
      "tokenId": "228",
      "type": "ERC721",
    },
  ],
  "txHash": "0xc814cb6b61222beda8a9bcc359e776e72ae732a1c29df572099ecef27c8461e4",
  "userAddress": "0x83eb387c8f7c8903f2be1ca367197fc4b3cb573e",
  "userName": "0x83eb",
}
`;

exports[`Interpreter 9`] = `
Object {
  "action": "repaid",
  "chainSymbol": "ETH",
  "contractAddress": "0x7d2768de32b0b80b7a3454c06bdac94a69ddc7a9",
  "contractName": "Aave V2",
  "counterpartyName": null,
  "exampleDescription": "0xd8a3 repaid 314305817611832341382659 CRV to Aave V2",
  "extra": Object {
    "tokenAmount": "314305817611832341382659",
    "tokenSymbol": "CRV",
  },
  "gasPaid": "0.012144212798965408",
  "nativeValueReceived": "0",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1656357349,
  "tokensReceived": Array [],
  "tokensSent": Array [
    Object {
      "address": "0x00ad8ebf64f141f1c81e9f8f792d3d1631c6c684",
      "amount": "314305817611832341382659",
      "name": "Aave variable debt bearing CRV",
      "symbol": "variableDebtCRV",
      "type": "ERC20",
    },
    Object {
      "address": "0xd533a949740bb3306d119cc777fa900ba034cd52",
      "name": "Curve DAO Token",
      "symbol": "CRV",
      "type": "unknown",
    },
  ],
  "txHash": "0xdf2f782ab0296121318cca140ef069f9f074c51ff4b11f0c677bcb01126f81de",
  "userAddress": "0xd8a3785869f3fc0f0019ec8e9f649b4c43a33383",
  "userName": "0xd8a3",
}
`;

exports[`Interpreter 10`] = `
Object {
  "action": "deposited",
  "chainSymbol": "ETH",
  "contractAddress": "0x7d2768de32b0b80b7a3454c06bdac94a69ddc7a9",
  "contractName": "Aave V2",
  "counterpartyName": null,
  "exampleDescription": "0x88a9 deposited 2800000000 USDT into Aave V2",
  "extra": Object {
    "tokenAmount": "2800000000",
    "tokenSymbol": "USDT",
  },
  "gasPaid": "0.015157096544929638",
  "nativeValueReceived": "0",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1656358062,
  "tokensReceived": Array [
    Object {
      "address": "0x3ed3b47dd13ec9a98b44e6204a523e766b225811",
      "amount": "2800000000",
      "name": "Aave interest bearing USDT",
      "symbol": "aUSDT",
      "type": "ERC20",
    },
  ],
  "tokensSent": Array [
    Object {
      "address": "0xdac17f958d2ee523a2206206994597c13d831ec7",
      "amount": "2800000000",
      "name": "Tether USD",
      "symbol": "USDT",
      "type": "ERC20",
    },
  ],
  "txHash": "0x24ee705da17a6061091880f47335d92950c72398980e271cdb9c69e8502827f4",
  "userAddress": "0x88a902adc0ff5bfe283d8d2365b51de6bc5e1666",
  "userName": "0x88a9",
}
`;

exports[`Interpreter 11`] = `
Object {
  "action": "swapped",
  "chainSymbol": "ETH",
  "contractAddress": "0x68b3465833fb72a70ecdf485e0e4c7bd8665fc45",
  "contractName": "Uniswap V3",
  "counterpartyName": null,
  "exampleDescription": "0x5596 swapped 280376616698 USDC for 150286887655178993883 WETH on Uniswap V3",
  "extra": Object {
    "tokenAmount0": "280376616698",
    "tokenAmount1": "150286887655178993883",
    "tokenName0": "USDC",
    "tokenName1": "WETH",
  },
  "gasPaid": "0.07226057005959764",
  "nativeValueReceived": "0",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1654532279,
  "tokensReceived": Array [
    Object {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "amount": "150286887655178993883",
      "name": "Wrapped Ether",
      "symbol": "WETH",
      "type": "unknown",
    },
  ],
  "tokensSent": Array [
    Object {
      "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
      "amount": "280376616698",
      "name": "USD Coin",
      "symbol": "USDC",
      "type": "ERC20",
    },
  ],
  "txHash": "0xe4737f2b6174dfcc9482a50aacdc0a919f12c14ede767d6dbfc2cd502747e4f5",
  "userAddress": "0x55961e9fc41c07c13ec0c7103b3d34c0ae32c2eb",
  "userName": "0x5596",
}
`;

exports[`Interpreter 12`] = `
Object {
  "action": "swapped",
  "chainSymbol": "ETH",
  "contractAddress": "0xe592427a0aece92de3edee1f18e0157c05861564",
  "contractName": "Uniswap V3",
  "counterpartyName": null,
  "exampleDescription": "0x7691 swapped 115000000000000000000 BYTES for 189236064 USDC on Uniswap V3",
  "extra": Object {
    "tokenAmount0": "115000000000000000000",
    "tokenAmount1": "189236064",
    "tokenName0": "BYTES",
    "tokenName1": "USDC",
  },
  "gasPaid": "0.01376121884064768",
  "nativeValueReceived": "0",
  "nativeValueSent": "0",
  "reverted": null,
  "timestamp": 1656530683,
  "tokensReceived": Array [
    Object {
      "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
      "amount": "189236064",
      "name": "USD Coin",
      "symbol": "USDC",
      "type": "ERC20",
    },
  ],
  "tokensSent": Array [
    Object {
      "address": "0x7d647b1a0dcd5525e9c6b3d14be58f27674f8c95",
      "amount": "115000000000000000000",
      "name": "BYTES",
      "symbol": "BYTES",
      "type": "ERC20",
    },
  ],
  "txHash": "0xf41a45f5347db0d95bfac48325f8272dce56782574e91083e5025324928bc40c",
  "userAddress": "0x76914474c1e922aa1166b1d073373b3b1a2dab47",
  "userName": "0x7691",
}
`;
